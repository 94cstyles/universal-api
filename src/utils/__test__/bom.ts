import { sleep } from './util';

/**
 * 模拟 localStorage
 * @param globals
 */
export function mockLocalStorage(globals) {
  const localStorage = {
    _data: new Map<string, string>(),
    setItem: (key: string, value: any) => {
      localStorage._data.set(key, String(value));
    },
    getItem: (key: string): string => {
      return localStorage._data.get(key) ?? null;
    },
    removeItem: (key: string) => {
      localStorage._data.delete(key);
    },
    clear: () => {
      localStorage._data.clear();
    },
  };

  globals.localStorage = localStorage;
  Object.defineProperty(window, 'localStorage', {
    get() {
      return localStorage;
    },
  });
}

/**
 * 模拟 xhr 对象
 * @param globals
 */
export function mockXHR(globals) {
  const response = {
    status: 200,
    data: '',
    headers: '',
  };

  const mockOpen = jest.fn();
  const mockSend = jest.fn(async function (data) {
    await sleep(10);
    this.readyState = 4;
    this.status = response.status ?? 200;
    this.response = response.data || '';
    this.onreadystatechange?.();
    await sleep(10);
    this.onload?.();
  });
  const mockSetRequestHeader = jest.fn();
  globals.FormData = window.FormData;
  globals.XMLHttpRequest = class {
    open = mockOpen;
    send = mockSend;
    setRequestHeader = mockSetRequestHeader;
    getAllResponseHeaders = () => response.headers || '';
  };

  function setResponse(status = 200, data?: any, headers?: Record<string, string>) {
    response.status = status;
    response.data = data;
    if (headers) {
      response.headers = Object.keys(headers)
        .map((key) => {
          const value = headers[key];
          return `${key}: ${value}`;
        })
        .join('\r\n');
    } else {
      response.headers = '';
    }
  }

  return {
    mockOpen,
    mockSend,
    mockSetRequestHeader,
    setResponse,
  };
}
