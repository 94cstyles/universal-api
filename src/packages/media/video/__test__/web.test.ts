import { testWebAPI } from '@/utils/__test__/util';
import { mockUserChooseMedia } from '@/utils/__test__/bom';

testWebAPI('video', async (globals) => {
  const originalRandom = Math.random;
  let _r = 0;
  Math.random = () => (++_r) / 1e8; // mock random
  const { chooseMedia, chooseVideo } = require('../src/index');

  chooseMedia();
  expect(document.body).toMatchSnapshot();
  await mockUserChooseMedia();
  expect(document.body).toMatchSnapshot();

  chooseVideo();
  expect(document.body).toMatchSnapshot();
  await mockUserChooseMedia();
  expect(document.body).toMatchSnapshot();

  // eslint-disable-next-line require-atomic-updates
  Math.random = originalRandom;
});
