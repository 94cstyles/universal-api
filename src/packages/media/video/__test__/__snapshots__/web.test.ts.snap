// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: video Test container: web 1`] = `
<body>
  <input
    id="input-1"
    multiple="multiple"
    name="file"
    style="display: none;"
    type="file"
  />
</body>
`;

exports[`Test Platform API: video Test container: web 2`] = `<body />`;

exports[`Test Platform API: video Test container: web 3`] = `
<body>
  <input
    accept="video/*"
    id="input-2"
    name="file"
    style="display: none;"
    type="file"
  />
</body>
`;

exports[`Test Platform API: video Test container: web 4`] = `<body />`;
